File: src/robot_stack/robot_stack/lidar_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Float32MultiArray


class LidarNode(Node):
    def __init__(self):
        super().__init__('lidar_node')
        self.declare_parameter('scan_topic', '/scan')
        self.declare_parameter('num_sectors', 4)
        self.declare_parameter('obstacle_distance', 0.5)

        scan_topic = self.get_parameter('scan_topic').value
        self.num_sectors = self.get_parameter('num_sectors').value
        self.obstacle_distance = self.get_parameter('obstacle_distance').value

        self.pub = self.create_publisher(Float32MultiArray, '/robot/obstacle', 10)
        self.sub = self.create_subscription(LaserScan, scan_topic, self.scan_callback, 10)
        self.get_logger().info('lidar_node started')

    def scan_callback(self, msg):
        if len(msg.ranges) == 0:
            return

        sector_size = len(msg.ranges) // self.num_sectors
        min_distances = []

        for i in range(self.num_sectors):
            start = i * sector_size
            end = start + sector_size if i < self.num_sectors - 1 else len(msg.ranges)
            sector_ranges = [r for r in msg.ranges[start:end] if r > msg.range_min and r < msg.range_max]
            min_dist = min(sector_ranges) if sector_ranges else msg.range_max
            min_distances.append(min_dist)

        # sectors: 0=front, 1=right, 2=back, 3=left
        front, right, back, left = min_distances[0], min_distances[1], min_distances[2], min_distances[3]

        obstacle_msg = Float32MultiArray()
        obstacle_msg.data = [front, right, back, left]
        self.pub.publish(obstacle_msg)

        # log if obstacle too close
        if front < self.obstacle_distance:
            self.get_logger().warn(f'Obstacle front: {front:.2f}m')


def main(args=None):
    rclpy.init(args=args)
    node = LidarNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/robot_stack/src/lidar_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>

class LidarNode : public rclcpp::Node
{
public:
  LidarNode() : Node("lidar_node")
  {
    this->declare_parameter("scan_topic", "/scan");
    this->declare_parameter("num_sectors", 4);
    this->declare_parameter("obstacle_distance", 0.5);

    scan_topic_ = this->get_parameter("scan_topic").as_string();
    num_sectors_ = this->get_parameter("num_sectors").as_int();
    obstacle_distance_ = this->get_parameter("obstacle_distance").as_double();

    pub_ = this->create_publisher<std_msgs::msg::Float32MultiArray>("/robot/obstacle", 10);
    sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
      scan_topic_, 10, std::bind(&LidarNode::scan_callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "lidar_node started");
  }

private:
  void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
  {
    size_t n = msg->ranges.size();
    if (n == 0) return;

    size_t sector_size = n / num_sectors_;
    std::vector<float> min_distances(num_sectors_, msg->range_max);

    for (int i = 0; i < num_sectors_; ++i) {
      size_t start = i * sector_size;
      size_t end = (i == num_sectors_ - 1) ? n : start + sector_size;
      for (size_t j = start; j < end; ++j) {
        float r = msg->ranges[j];
        if (r > msg->range_min && r < msg->range_max && r < min_distances[i]) {
          min_distances[i] = r;
        }
      }
    }

    float front = min_distances[0];
    float right = min_distances[1];
    float back = min_distances[2];
    float left = min_distances[3];

    auto obstacle_msg = std_msgs::msg::Float32MultiArray();
    obstacle_msg.data = {front, right, back, left};
    pub_->publish(obstacle_msg);

    if (front < obstacle_distance_) {
      RCLCPP_WARN(this->get_logger(), "Obstacle front: %.2f m", front);
    }
  }

  std::string scan_topic_;
  int num_sectors_;
  double obstacle_distance_;
  rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pub_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<LidarNode>());
  rclcpp::shutdown();
  return 0;
}
